File: ios/dataProvider/repositories/FileRepository.m
//
//  FileRepository.m
//  StorjMobile
//
//  Created by Barterio on 3/23/18.
//  Copyright © 2018 Storj. All rights reserved.
//

#import "FileRepository.h"
#import "FileDbo.h"
#import "FileModel.h"
#import "FileContract.h"

#import "DictionaryUtils.h"

#import "Response.h"

@implementation FileRepository

static NSArray * columns;

-(instancetype) init
  {
    return [super init];
  }

-(NSArray *) getAll {
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME];
  __block NSMutableArray<FileDbo *> * fileDboArray = [NSMutableArray<FileDbo *> array];
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet * resultSet = [db executeQuery:request];
    if(!resultSet) {
      NSLog(@"No result set returned");
      
      return;
    }
    
    while ([resultSet next]) {
      FileDbo * dbo = [FileRepository getFileDboFromResultSet:resultSet];
      if(dbo) {
        [fileDboArray addObject:dbo];
      }
    }
    [resultSet close];
  }];
  [queue close];
  
  return fileDboArray;
}

-(NSArray *) getAllFromBucket:(NSString *)bucketId {
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ WHERE %@ = ?",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME,
                       FileContract.FILE_FK];
  __block NSMutableArray <FileDbo *> * fileDboArray = [NSMutableArray <FileDbo *> array];
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet *resultSet = [db executeQuery:request, bucketId];
    if(!resultSet) {
      NSLog(@"No result set returned");
      return;
    }
    
    while([resultSet next]) {
      FileDbo *dbo = [FileRepository getFileDboFromResultSet:resultSet];
      if(dbo){
        [fileDboArray addObject:dbo];
      }
    }
    [resultSet close];
  }];
  [queue close];
  
  return fileDboArray;
}

-(NSArray *) getAllFromBucket:(NSString *)bucketId
                orderByColumn: (NSString *) columnName
                   descending: (BOOL) isDescending {
  NSString *orderByColumn = columnName;
  if(!columnName || columnName.length == 0) {
    orderByColumn = FileContract.NAME;
  }
  
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ WHERE %@ = ? ORDER BY %@ COLLATE NOCASE %@",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME,
                       FileContract.FILE_FK,
                       orderByColumn,
                       isDescending ? @"DESC" : @"ASC"];
  __block NSMutableArray <FileDbo *> * fileDboArray = [NSMutableArray <FileDbo *> array];
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet *resultSet = [db executeQuery:request, bucketId];
    if(!resultSet) {
      NSLog(@"No result set returned");
      return;
    }
    
    while([resultSet next]) {
      FileDbo *dbo = [FileRepository getFileDboFromResultSet:resultSet];
      if(dbo){
        [fileDboArray addObject:dbo];
      }
    }
    [resultSet close];
  }];
  [queue close];
  
  return fileDboArray;
}

-(NSArray *) getAllWithOrderByColumn: (NSString *) columnName
                               order: (BOOL) isDescending {
  NSString *orderByColumn = columnName;
  if(!columnName || [columnName length] == 0) {
    orderByColumn = FileContract.CREATED;
  }
  
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ ORDER BY %@ COLLATE NOCASE s%@",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME,
                       orderByColumn,
                       isDescending ? @"DESC" : @"ASC"];
  __block NSMutableArray<FileDbo *> * fileDboArray = [NSMutableArray<FileDbo *> array];
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet * resultSet = [db executeQuery:request];
    if(!resultSet) {
      NSLog(@"No result set returned");
  
      return;
    }
    
    while ([resultSet next]) {
      FileDbo * dbo = [FileRepository getFileDboFromResultSet:resultSet];
      if(dbo) {
        [fileDboArray addObject:dbo];
      }
    }
    [resultSet close];
  }];
  [queue close];
  
  return fileDboArray;
}

-(NSArray *) getStarred {
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ WHERE %@ = ?",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME,
                       FileContract.STARRED];
  __block NSMutableArray<FileDbo *> * fileDboArray = [NSMutableArray<FileDbo *> array];
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet * resultSet = [db executeQuery:request, 1];
    if(!resultSet) {
      NSLog(@"No result set returned");
      
      return;
    }
  
    while ([resultSet next]) {
      FileDbo * dbo = [FileRepository getFileDboFromResultSet:resultSet];
      if(dbo) {
        [fileDboArray addObject:dbo];
      }
    }
    [resultSet close];
  }];
  [queue close];
  
  return fileDboArray;
}

-(FileDbo *) getByFileId:(NSString *) fileId {
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ WHERE %@ = ?",
                       [[FileRepository getSelectionColumnsString]componentsJoinedByString:@","],
                       FileContract.TABLE_NAME,
                       FileContract.FILE_ID];
  __block FileDbo * dbo = nil;
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet * resultSet = [db executeQuery:request, fileId];
    if(!resultSet) {
      NSLog(@"No result set returned");
      return;
    }
    
    if([resultSet next]) {
      dbo = [FileRepository getFileDboFromResultSet:resultSet];
    }
    [resultSet close];
  }];
  [queue close];
  
  return dbo;
}

-(FileDbo *) getByColumnName:(NSString *) columnName
                 columnValue:(NSString *) columnValue {
  NSString *request = [NSString stringWithFormat:@"SELECT %@ FROM %@ WHERE %@ = ?",
                       columnName,
                       FileContract.TABLE_NAME,
                       columnName];
  __block FileDbo *dbo = nil;
  FMDatabaseQueue *queue = [self readableQueue];
  [queue inDatabase:^(FMDatabase * _Nonnull db) {
    FMResultSet * resultSet = [db executeQuery:request, columnValue];
    if(!resultSet) {
      NSLog(@"No result set returned");
      return;
    }
    
    if([resultSet next]){
      dbo = [FileRepository getFileDboFromResultSet:resultSet];
    }
    [resultSet close];
  }];
  [queue close];
  
  return dbo;
}

-(Response *) insertWithModel: (FileModel *) model {
  NSLog(@"Inserting file: %@", [model toDictionary]);
  if(!model || ![model isValid]) {
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeInsertIntoTable:FileContract.TABLE_NAME
                              fromDict:[[FileDbo fileDboFromFileModel:model] toDictionary]];
}

-(Response *) deleteByModel: (FileModel *) model {
  if(!model || ![model isValid]) {
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeDeleteFromTable:FileContract.TABLE_NAME
                         withObjectKey:FileContract.FILE_ID
                      withObjecktValue:[model _fileId]];
}

-(Response *) deleteById: (NSString *) fileId {
  if(!fileId || [fileId length] == 0) {
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeDeleteFromTable:FileContract.TABLE_NAME
                         withObjectKey:FileContract.FILE_ID
                      withObjecktValue:fileId];
}

-(Response *) deleteByIds: (NSArray *) fileIds {
  if(!fileIds || [fileIds count] == 0) {
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeDeleteFromTable:FileContract.TABLE_NAME
                         withObjectKey:FileContract.FILE_ID
                        withObjecktIds:fileIds];
}

-(Response *) deleteAll {
  
   return [super executeDeleteAllFromTable:FileContract.TABLE_NAME];
}

-(Response *) deleteAllFromBucket:(NSString *) bucketId {
  
  return [super executeDeleteFromTable:FileContract.TABLE_NAME
                         withObjectKey:FileContract.FILE_FK
                         withObjecktValue:bucketId];
}

-(Response *) updateByModel: (FileModel *) model {
  if(!model || ![model isValid]) {
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  NSMutableDictionary *updateDictionary = [NSMutableDictionary dictionary];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _created]]
                       forKey:FileContract.CREATED];
  [updateDictionary setObject:@([model _isDecrypted])
                       forKey:FileContract.DECRYPTED];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _erasure]]
                       forKey:FileContract.ERASURE];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _hmac]]
                       forKey:FileContract.HMAC];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _index]]
                       forKey:FileContract.INDEX];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _mimeType]]
                       forKey:FileContract.MIME_TYPE];
  [updateDictionary setObject:@([model _size])
                       forKey:FileContract.SIZE];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _bucketId]]
                       forKey:FileContract.FILE_FK];
  [updateDictionary setObject:[DictionaryUtils checkAndReturnNSString:[model _name]]
                       forKey:FileContract.NAME];
  
  return [super executeUpdateAtTable:FileContract.TABLE_NAME
                           objectKey:FileContract.FILE_ID
                            objectId:[model _fileId]
                    updateDictionary:updateDictionary];
}

-(Response *) updateById:(NSString *)fileId
                 starred:(BOOL) isStarred {
  
  if(!fileId || [fileId length] == 0) {
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeUpdateAtTable:FileContract.TABLE_NAME
                           objectKey:FileContract.FILE_ID
                            objectId:fileId
                    updateDictionary:@{FileContract.STARRED:@(isStarred)}];
}
-(Response *) updateById: (NSString *) fileId
           downloadState: (int) downloadState
              fileHandle: (long) fileHandle
                 fileUri: (NSString *) fileUri {
  NSLog(@"fileId %@, downloadState %d, fileHandle %lu, uri %@", fileId, downloadState, fileHandle, fileUri);
  
  if(!fileId || fileId.length == 0){
    
    return [Response errorResponseWithMessage:@"Model is not valid"];
  }
  
  return [super executeUpdateAtTable:FileContract.TABLE_NAME
                           objectKey:FileContract.FILE_ID
                            objectId:[DictionaryUtils checkAndReturnNSString:fileId]
                    updateDictionary:@{FileContract.DOWNLOAD_STATE: @(downloadState),
                                       FileContract.FILE_HANDLE: @(fileHandle),
                                       FileContract.FILE_URI: [DictionaryUtils checkAndReturnNSString:fileUri]}];
}

-(Response *) updateThumbnailWithFileId: (NSString *) fileId
                        thumbnailBase64: (NSString *) thumbnailBase64String {
  if(!fileId || fileId.length == 0){
    return [Response errorResponseWithMessage:@"FileId is not valid"];
  }
  
  if(!thumbnailBase64String || thumbnailBase64String.length == 0){
    return [Response errorResponseWithMessage:@"Base64 representation of image is not valid"];
  }
  
  return [super executeUpdateAtTable:FileContract.TABLE_NAME
                           objectKey:FileContract.FILE_ID
                            objectId:fileId
                    updateDictionary:@{FileContract.FILE_THUMBNAIL: thumbnailBase64String}];
}

+(FileDbo *) getFileDboFromResultSet:(FMResultSet *) resultSet {
  
  return [[FileDbo alloc] initWithBucketId:[resultSet stringForColumn:FileContract.FILE_FK]
                                   created:[resultSet stringForColumn:FileContract.CREATED]
                                   erasure:[resultSet stringForColumn:FileContract.ERASURE]
                                      hmac:[resultSet stringForColumn:FileContract.HMAC]
                                    fileId:[resultSet stringForColumn:FileContract.FILE_ID]
                                     index:[resultSet stringForColumn:FileContract.INDEX]
                                  mimeType:[resultSet stringForColumn:FileContract.MIME_TYPE]
                                      name:[resultSet stringForColumn:FileContract.NAME]
                                      size:[resultSet longForColumn:FileContract.SIZE]
                               isDecrypted:[resultSet boolForColumn:FileContract.DECRYPTED]
                                 isStarred:[resultSet boolForColumn:FileContract.STARRED]
                                  isSynced:NO
                             downloadState:[resultSet intForColumn:FileContract.DOWNLOAD_STATE]
                                fileHandle:[resultSet longForColumn:FileContract.FILE_HANDLE]
                                   fileUri:[resultSet stringForColumn:FileContract.FILE_URI]
                                 thumbnail:[resultSet stringForColumn:FileContract.FILE_THUMBNAIL]];
}

+(NSArray *)getSelectionColumnsString {
  if(!columns) {
    NSMutableArray *colArray = [NSMutableArray array];
//    [colArray addObject:FileContract.ID];
    [colArray addObject:FileContract.FILE_ID];
    [colArray addObject:FileContract.NAME];
    [colArray addObject:FileContract.MIME_TYPE];
    [colArray addObject:FileContract.INDEX];
    [colArray addObject:FileContract.HMAC];
    [colArray addObject:FileContract.ERASURE];
    [colArray addObject:FileContract.CREATED];
    [colArray addObject:FileContract.DECRYPTED];
    [colArray addObject:FileContract.SIZE];
    [colArray addObject:FileContract.STARRED];
    [colArray addObject:FileContract.SYNCED];
    [colArray addObject:FileContract.FILE_FK];
    [colArray addObject:FileContract.DOWNLOAD_STATE];
    [colArray addObject:FileContract.FILE_HANDLE];
    [colArray addObject:FileContract.FILE_URI];
    [colArray addObject:FileContract.FILE_THUMBNAIL];
    columns = [NSArray arrayWithArray:colArray];
  }
  
  return columns;
}

@end

File: android/app/src/main/java/io/storj/mobile/storjlibmodule/dataprovider/repositories/FileRepository.java
package io.storj.mobile.storjlibmodule.dataprovider.repositories;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.storj.mobile.storjlibmodule.models.FileModel;
import io.storj.mobile.storjlibmodule.responses.Response;
import io.storj.mobile.storjlibmodule.dataprovider.dbo.FileDbo;
import io.storj.mobile.storjlibmodule.dataprovider.contracts.FileContract;

/**
 * Created by crawt on 3/4/2018.
 */

public class FileRepository extends BaseRepository {
    private String[] _columns = new String[] {
        FileContract._ID,
        FileContract._CREATED,
        FileContract._DECRYPTED,
        FileContract._ERASURE,
        FileContract._HMAC,
        FileContract._INDEX,
        FileContract._MIMETYPE,
        FileContract._STARRED,
        FileContract._SIZE,
        FileContract._SYNCED,
        FileContract._DOWNLOAD_STATE,
        FileContract._FILE_HANDLE,
        FileContract.FILE_FK,
        FileContract._NAME,
        FileContract._FILE_URI,
        FileContract._FILE_THUMBNAIL
    };

    @Inject
    public FileRepository(SQLiteDatabase db) { super(db); }

    public List<FileDbo> getAll() {
        List<FileDbo> result = new ArrayList();
        Cursor cursor = _db.query(FileContract.TABLE_NAME, null, null, null, null, null, null, null);

        result = _getListFromCursor(cursor);

        cursor.close();

        return result;
    }

    public List<FileDbo> getAll(String orderByColumn, boolean isDesc) {
        List<FileDbo> result = new ArrayList();

        String column = orderByColumn;

        if(orderByColumn == null || orderByColumn.isEmpty()) {
            column = FileContract._NAME;
        }

        String orderBy = isDesc ? column + " DESC" : orderByColumn + " ASC";

        Cursor cursor = _db.query(FileContract.TABLE_NAME, null, null, null, null, null, orderBy, null);

        result = _getListFromCursor(cursor);

        cursor.close();

        return result;
    }

    public List<FileDbo> getAll(String bucketId) {
        List<FileDbo> result = new ArrayList();
        String[] selectionArgs = {
                bucketId
        };
        Cursor cursor = _db.query(FileContract.TABLE_NAME,
                null,
                FileContract.FILE_FK + " = ?",
                selectionArgs,
                null,
                null,
                null,
                null);

        result = _getListFromCursor(cursor);

        cursor.close();

        return result;
    }

    public List<FileDbo> getAll(String bucketId, String orderByColumn, boolean isDesc) {
        List<FileDbo> result = new ArrayList();
        String[] selectionArgs = {
                bucketId
        };

        String column = orderByColumn;

        if(orderByColumn == null || orderByColumn.isEmpty()) {
            column = FileContract._NAME;
        }

        String orderBy = isDesc ? column + " DESC" : column + " ASC";

        Cursor cursor = _db.query(FileContract.TABLE_NAME,
                null,
                FileContract.FILE_FK + " = ?",
                selectionArgs,
                null,
                null,
                orderBy,
                null);

        result = _getListFromCursor(cursor);

        cursor.close();

        return result;
    }

    public List<FileDbo> getAllCollateNocase(String bucketId, String orderByColumn, boolean isDesc) {
        List<FileDbo> result = new ArrayList();
        String[] selectionArgs = {
                bucketId
        };

        String column = orderByColumn;

        if(orderByColumn == null || orderByColumn.isEmpty()) {
            column = FileContract._NAME;
        }

        String orderBy = isDesc ? " ORDER BY " + column + " COLLATE NOCASE DESC;" : " ORDER BY " + column + " COLLATE NOCASE ASC;";

        String query = "SELECT * FROM " + FileContract.TABLE_NAME + " WHERE " + FileContract.FILE_FK + " = ?" + orderBy;

        Cursor cursor = _db.rawQuery(query, new String[] { bucketId  });

        result = _getListFromCursor(cursor);

        cursor.close();

        return result;
    }

    public FileDbo get(String fileId) {
        FileDbo model = null;
        String[] selectionArgs = {
                fileId
        };
        String orderBy = FileContract._CREATED + " DESC";

        Cursor cursor = _db.query(
                FileContract.TABLE_NAME,
                null,
                FileContract._ID + " = ?",
                selectionArgs,
                null, null, orderBy, null);

        model = _getSingleFromCursor(cursor);

        cursor.close();

        return model;
    }

    public FileDbo get(String param, String selection) {
        FileDbo model = null;
        String[] selectionArgs = {
                param
        };
        String orderBy = FileContract._CREATED + " DESC";

        Cursor cursor = _db.query(
                FileContract.TABLE_NAME,
                null,
                selection + " = ?",
                selectionArgs,
                null, null, orderBy, null);

        model = _getSingleFromCursor(cursor);

        cursor.close();

        return model;
    }

    public FileDbo get(String param, String selection, String bucketId) {
        FileDbo model = null;
        String[] selectionArgs = {
                param,
                bucketId
        };
        String orderBy = FileContract._CREATED + " DESC";

        Cursor cursor = _db.query(
                FileContract.TABLE_NAME,
                null,
                selection + " = ? and " + FileContract.FILE_FK + " = ?",
                selectionArgs,
                null, null, orderBy, null);

        model = _getSingleFromCursor(cursor);

        cursor.close();

        return model;
    }

    public Response insert(FileModel model) {
        if(model == null || !model.isValid())
            return new Response(false, "Model is not valid!");

        ContentValues map = new ContentValues();

        map.put(FileContract._ID, model.getFileId());
        map.put(FileContract._CREATED, model.getCreated());
        map.put(FileContract._DECRYPTED, model.getDecrypted());
        map.put(FileContract._ERASURE, model.getErasure());
        map.put(FileContract._HMAC, model.getHmac());
        map.put(FileContract._INDEX, model.getIndex());
        map.put(FileContract._MIMETYPE, model.getMimeType());
        map.put(FileContract._STARRED, model.getStarred());
        map.put(FileContract._SYNCED, model.isSynced());
        map.put(FileContract._DOWNLOAD_STATE, model.downloadState());
        map.put(FileContract._FILE_HANDLE, model.getFileHandle());
        map.put(FileContract._FILE_URI, model.getUri());
        map.put(FileContract._SIZE, model.getSize());
        map.put(FileContract.FILE_FK, model.getBucketId());
        map.put(FileContract._NAME, model.getName());
        map.put(FileContract._FILE_THUMBNAIL, model.getThumbnail());

        return _executeInsert(FileContract.TABLE_NAME, map);
    }

    public Response delete(FileModel model) {
        if(model == null || !model.isValid())
            return new Response(false, "Model is not valid!");

        return _executeDelete(new String[] { model.getFileId() }, FileContract.TABLE_NAME, FileContract._DEFAULT_WHERE_CLAUSE);
    }

    public Response delete(String fileId) {
        if(fileId == null || fileId.isEmpty())
            return new Response(false, "Model id is not valid!");

        return _executeDelete(new String[] { fileId }, FileContract.TABLE_NAME, FileContract._DEFAULT_WHERE_CLAUSE);
    }

    public Response delete(String[] fileIdList) {
        if(fileIdList == null || fileIdList.length == 0)
            return new Response(false, "Model list is not valid!");

        return _executeDelete(fileIdList, FileContract.TABLE_NAME, FileContract._DEFAULT_WHERE_CLAUSE);
    }

    public Response deleteAll(String bucketId) {
        return _deleteAll(FileContract.TABLE_NAME);
    }

    public Response update(FileModel model) {
        if(model == null || !model.isValid())
            return new Response(false, "Model is not valid!");

        ContentValues map = new ContentValues();

        map.put(FileContract._CREATED, model.getCreated());
        map.put(FileContract._DECRYPTED, model.getDecrypted());
        map.put(FileContract._ERASURE, model.getErasure());
        map.put(FileContract._HMAC, model.getHmac());
        map.put(FileContract._INDEX, model.getIndex());
        map.put(FileContract._MIMETYPE, model.getMimeType());
        map.put(FileContract._SIZE, model.getSize());
        map.put(FileContract.FILE_FK, model.getBucketId());
        map.put(FileContract._NAME, model.getName());

        return _executeUpdate(FileContract.TABLE_NAME, model.getFileId(), null,null, map);
    }

    /**
     * Updates isStarred field
     * @param fileId id of the file to update
     * @param isStarred indicates state of isStarred filed
     */
    public Response update(String fileId, boolean isStarred) {
        if(fileId == null || fileId.isEmpty())
            return new Response(false, "File id is not valid!");

        ContentValues map = new ContentValues();

        map.put(FileContract._STARRED, isStarred ? 1 : 0);

        return _executeUpdate(FileContract.TABLE_NAME, fileId, null,null, map);
    }

    public Response update(String fileId, int downloadState, long fileHandle, String fileUri) {
        if(fileId == null || fileId.isEmpty())
            return new Response(false, "File id is not valid!");

        ContentValues map = new ContentValues();

        map.put(FileContract._DOWNLOAD_STATE, downloadState);
        map.put(FileContract._FILE_HANDLE, fileHandle);
        map.put(FileContract._FILE_URI, fileUri);

        return _executeUpdate(FileContract.TABLE_NAME, fileId, null,null, map);
    }

    /**
     * Updates thumbnail
     * @param fileId id of the file to update
     * @param thumbnailBase64String converted to base64 string thumbnail
     */
    public Response updateThumbnail(String fileId, String thumbnailBase64String) {
        if(fileId == null || fileId.isEmpty())
            return new Response(false, "File id is not valid!");

        ContentValues map = new ContentValues();

        map.put(FileContract._FILE_THUMBNAIL, thumbnailBase64String);

        return _executeUpdate(FileContract.TABLE_NAME, fileId, null,null, map);
    }

    private FileDbo _getSingleFromCursor(Cursor cursor) {
        FileDbo model = null;

        if (cursor.moveToFirst()){
            model = _fillFile(cursor);
        }

        return model;
    }


    private List<FileDbo> _getListFromCursor(Cursor cursor) {
        List<FileDbo> result = new ArrayList();

        if (cursor.moveToFirst()){
            do {
                result.add(_fillFile(cursor));
            } while (cursor.moveToNext());
        }

        return result;
    }

    private FileDbo _fillFile(Cursor cursor) {
        FileDbo model = new FileDbo();

            for(int i = 0; i < _columns.length; i++) {
                switch(_columns[i]) {
                    case FileContract._CREATED :
                    case FileContract._NAME :
                    case FileContract._ID :
                    case FileContract._ERASURE:
                    case FileContract._HMAC:
                    case FileContract._INDEX:
                    case FileContract._MIMETYPE:
                    case FileContract.FILE_FK:
                    case FileContract._FILE_URI:
                    case FileContract._FILE_THUMBNAIL:
                        model.setProp(_columns[i], cursor.getString(cursor.getColumnIndex(_columns[i])));
                        break;
                    case FileContract._DECRYPTED :
                    case FileContract._STARRED :
                    case FileContract._SYNCED :
                        model.setProp(_columns[i], cursor.getInt(cursor.getColumnIndex(_columns[i])) == 1 ? true : false);
                        break;
                    case FileContract._DOWNLOAD_STATE:
                        model.setProp(_columns[i], cursor.getInt(cursor.getColumnIndex(_columns[i])));
                        break;
                    case FileContract._SIZE :
                    case FileContract._FILE_HANDLE:
                        model.setProp(_columns[i], cursor.getLong(cursor.getColumnIndex(_columns[i])));
                        break;
                }
            }

        return model;
    }
}
